### ISSUE TITLE
manifest list (multiarch) picks wrong arch on ARMv6

### ISSUE THREAD

**Issue by kremerol** [moby/moby#41017]
**Description**

The command `docker pull` executed on a armv6 Raspberry Pi One ("Model B Plus Rev 1.2") running a 32-bit Raspbian OS will fail to pull the correct docker image 
defined in a manifest as soon as a 'linux/arm/v7' manifest entry exists.

**Steps to reproduce the issue:**
1. `uname -m && docker run --rm arm32v6/alpine:3.8 sh -c 'cat /etc/*release | grep PRETTY_NAME'`
2. `uname -m && docker run --rm arm32v6/alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME'`
3. `uname -m && docker run --rm alpine:3.8 sh -c 'cat /etc/*release | grep PRETTY_NAME'`
4. `uname -m && docker run --rm alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME'`

**Describe the results you received:**
```
$ uname -m && docker run --rm arm32v6/alpine:3.8 sh -c 'cat /etc/*release | grep PRETTY_NAME'
armv6l
PRETTY_NAME="Alpine Linux v3.8"
$ uname -m && docker run --rm arm32v6/alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME'
armv6l
PRETTY_NAME="Alpine Linux v3.9"
$ uname -m && docker run --rm alpine:3.8 sh -c 'cat /etc/*release | grep PRETTY_NAME'
armv6l
PRETTY_NAME="Alpine Linux v3.8"
$ uname -m && docker run --rm alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME'
armv6l
$ echo $?
139
```
For steps 1,2 and 3, I was expecting these results. However, at step 4, the 'docker run' command did not produce any output and exits 
with a 139 return code.

**Describe the results you expected:**

For step 4, I was expecting the same result as in step 2: 
```
armv6l
PRETTY_NAME="Alpine Linux v3.9"
```

**Additional information you deem important (e.g. issue happens only occasionally):**

There is no "linux/arm/v7" image on alpine:3.8 manifest:
```
$ docker run --rm mplatform/mquery alpine:3.8Image: alpine:3.8
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
```

But alpine:3.9 manifest defines a linux/arm/v7 entry:
```
$ docker run --rm mplatform/mquery alpine:3.9
Image: alpine:3.9
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm/v7
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
```

If you prefer, with `docker manifest inspect`:
```
$ DOCKER_CLI_EXPERIMENTAL=enabled docker manifest inspect alpine:3.8
[...]
"digest": "sha256:dabea2944dcc2b86482b4f0b0fb62da80e0673e900c46c0e03b45919881a5d84",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v6"
         }
[...]

$ DOCKER_CLI_EXPERIMENTAL=enabled docker manifest inspect alpine:3.9
[...]
{
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:7a3d88cbc7e2d6c0213deaf2d006933c9f5905c4eb7846b703a66fc6504000b7",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v6"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 528,
         "digest": "sha256:cfd8b55d209956f63c8fcc931f5c6874984e5e0ffdcb8f45ba9085f190385d73",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
[...]
```

An evidence that dockerd 19.03.8 pulls the (wrong) linux/arm/v7 images variant:  
```
$ docker pull arm32v7/alpine:3.9
$ docker images | grep "alpine"
arm32v6/alpine        3.8                                                   e60f4bcd6c89        4 months ago        4.01MB    # OK
alpine                3.8                                                   e60f4bcd6c89        4 months ago        4.01MB    # OK
arm32v6/alpine        3.9                                                   4fe9c9b2a0c4        4 weeks ago         4.68MB    # OK
arm32v7/alpine        3.9                                                   9df0ff5446fc        4 weeks ago         3.72MB    # OK
alpine                3.9                                                   9df0ff5446fc        4 weeks ago         3.72MB    # NOK: should be '4fe9c9b2a0c4' and not '9df0ff5446fc' ! 
```

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:35:24 2020
 OS/Arch:           linux/arm
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:29:22 2020
  OS/Arch:          linux/arm
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**Output of `docker info`:**

```
Client:
 Debug Mode: false

Server:
 Containers: 2
  Running: 2
  Paused: 0
  Stopped: 0
 Images: 34
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: active
  NodeID: 2zsjxcgeymlaecu00y8j0i1uw
  Is Manager: true
  ClusterID: t5ftctfw0dlnxbve8ilm7yo0x
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8  
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.0.114
  Manager Addresses:
   192.168.0.114:2377
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.97+
 Operating System: Raspbian GNU/Linux 10 (buster)
 OSType: linux
 Architecture: armv6l
 CPUs: 1
 Total Memory: 432.4MiB
 Name: white
 ID: QDNQ:Y3WR:ALAX:BXDG:JRKM:VBXW:SEC3:LX7D:LLFH:GG2O:EHK6:GG4G
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
WARNING: No cpuset support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

This test was ran on a "Raspberry Pi Model B Plus Rev 1.2".
Important device info: it is an armhf v6 (and not v7 !) device, even if its 'CPU architecture' is set to 7 ! 
```
$ dpkg --print-architecture
armhf
$ uname -m
armv6l
$ cat /proc/cpuinfo | grep "CPU architecture"
CPU architecture: 7
```

**Comment by kremerol** [comment]
**My 2 cents analysis:** 
1. I know that the ARM world is very complex (see https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures and/or https://en.wikipedia.org/wiki/ARM_architecture), 
   and I know it is much much more complex than only 'armv32v5, arm32v6, arm32v7 or arm64v8'.
2. For debian, armhf is armv7 + hard floating-point (see https://wiki.debian.org/ArmHardFloatPort#Port_naming_debate_notes). But that's not the case for Raspbian !
   ```
   Raspbian builds a single image for all of the Raspberry families, so you will get an armhf 32-bit, hard floating-point system, but built for the ARMv6 ISA (with VFP2), unlike Debian's ARMv7 ISA (with VFP3) port.
   ``` 
   See https://wiki.debian.org/RaspberryPi#Debian_and_Raspbian.
3. On a "Raspberry Pi Model B Plus Rev 1.2" (as reported by `cat /proc/device-tree/model`), the `lscpu` command and 
   the `uname -m` (or `arch`) clearly understand it is an arm v6 device, while `dpkg --print-architecture` confirm it is a hard floating-point system:
   ```
   $ lscpu -J | grep Architecture
        {"field":"Architecture:", "data":"armv6l"},
   $ uname -m
   armv6l
   $ arch
   armv6l
   $ `dpkg --print-architecture`
   armhf
   ```
4. This arm32v6 (armv6l) device as a 'CPU Architecture' set by the kernel to 7 !
   ```
   $ cat /proc/cpuinfo 
   processor       : 0
   model name      : ARMv6-compatible processor rev 7 (v6l)
   BogoMIPS        : 697.95
   Features        : half thumb fastmult vfp edsp java tls 
   CPU implementer : 0x41
   CPU architecture: 7
   CPU variant     : 0x0
   CPU part        : 0xb76
   CPU revision    : 7
   
   Hardware        : BCM2835
   Revision        : 0010
   Serial          : 00000000c0c1d288
   Model           : Raspberry Pi Model B Plus Rev 1.2
   ```
5. By the way, this produce some discussions already in 2012 (see https://www.raspberrypi.org/forums/viewtopic.php?t=12614).
   Apparently, `CPU architecture` is set to 7 when the cpu support "VMSAv7 or PMSAv7", but it is not related to 'ARM v6 or v7'.
6. Notes that `dpkg --print-architecture` returns `armhf` for this device
7. Containerd use the "CPU architecture" field of the '/proc/cpuinfo' pseudo file to set the arm device 'variant':
   See https://github.com/containerd/containerd/blob/release/1.3/platforms/cpuinfo.go#L93.
   In our case, it is a mistake, because containerd will consider this RPI1 device as a 'arm32v7' device, while it is not true !
8. By the way, https://github.com/containerd/containerd/blob/release/1.3/platforms/database.go#L98 consider that 'armhf' is an equivalent to 'arm32v7',
   which is also not always true neither (it is at least not true for this 'Raspberry Pi Model B Plus Rev 1.2' running the Raspian OS)
9. Moby (and therefore Docker), embeds the [containerd code](https://github.com/moby/moby/blob/19.03/vendor/github.com/containerd/containerd/platforms/cpuinfo.go#L93) 
   and use it to determine the local server plateform: see https://github.com/moby/moby/blob/19.03/distribution/pull_v2_unix.go#L21.
10. When 'docker pull' needs to find the correct docker image according to the manifest, it uses this containerd code to find the
    [platforms.DefaultSpec()](https://github.com/moby/moby/blob/19.03/distribution/pull_v2_unix.go#L51) which will in our case return 'v7'.
11. Consequence: as soon as a 'arm32v7' image will be declared in the manifest, the [filterManifests](https://github.com/moby/moby/blob/19.03/distribution/pull_v2_unix.go#L26)
    method will be happy to find a perfect match that respects the [OCI specs](https://github.com/opencontainers/image-spec/blob/v1.0.1/image-index.md#image-index-property-descriptions), 
    and that's the image that will be downloaded. And in that case, it will fail as the device is not armv7 !
12. However, if there is not 'v7' variant in the manifest, the 'fallback' mode will be trigger, and all the 'arm' manifest entry will be added to the 'matches' list 
    (which is by the way CURRENTLY a little bit useless as the 'best matching' algorithm will simply return the first found variant (see https://github.com/moby/moby/blob/19.03/distribution/pull_v2.go#L792)
    so today, stopping on the first match could be enough. But's that not what I recommend, because the full list will be needed to fix the issue [Issue 37647](https://github.com/moby/moby/issues/37647)).

**Comment by kremerol** [comment]
I have found bugs report related to this one, but I think that none of these issues are really a duplicate of that one :   
     - [Issue 37647 - ARMv6 machine pulls v7 image from manifest list](https://github.com/moby/moby/issues/37647): the issue title is matching with this issue, but not the detailed description:        "If I pull a multi-platform image e.g. httpd:2.1.34 with a Raspberry Pi Zero (ARMv6 CPU) it pulls the v7 image it can't execute, *even though there is a v5 image in the manifest list*".
       So even if the 'v6/v7' variant detection will be fixed, this 'get v5 instead of v7' issue could remain. Sorting the 'arm maching list' (v5 first, v6 second, v7) before returning the first 
       element should solve this issue.
     - [Issue 34875 - manifest list (multiarch) picks wrong arch on ARMv7](https://github.com/moby/moby/issues/34875): for me, it is a duplicate of 37647, as it should be solved by ordering arm variants. 
       By the way, arm64v8 devices are compatible with arm32. So I wonder if the 'sortering algo' should not also consider this use case.

**Comment by kremerol** [comment]
**Additional test evidence with alpine:**
```
docker run -it --rm alpine:3.8 apk --version
# is working because arm32v7 manifest do not exist for alpine 3.8 !
# It returns: "apk-tools 2.10.1, compiled for armhf."
# => OK

docker run -it --rm alpine:3.11.6 apk --version 
# Is failing because arm32v7 manifest exists
# FYI, it returns "apk-tools 2.10.5, compiled for armv7." on real arm32v7 devices
# => NOK ! That's a bug !!!  

docker run -it --rm arm32v6/alpine:3.11.6 apk --version
# Is working, because we bypass manifest parsing and we get directly the correct image. 
# It returns: "apk-tools 2.10.5, compiled for armhf."
# => OK
```

**And with debian:**
```    
docker run --rm -it arm32v5/debian dpkg --print-architecture
# return "armel"
# => OK

docker run --rm -it arm32v6/debian dpkg --print-architecture
# fails, because "repository does not exist"
# => OK

docker run --rm -it arm32v7/debian dpkg --print-architecture
# 'fails with error code' because device is not armv7
# => OK

docker run --rm -it debian dpkg --print-architecture
# 'fails with error code'
# => NOK ! That's a bug !!!

# Notes that 'uname -a' or 'uname -m' is not a good test to get the arch inside a docker container. I guess that these commands query the kernel,
# so we get the architecture of the 'os running docker' and not the architecture of the 'container'.  
```

And by the way, upgrading to docker 19.03.9 did not solve the issue:
```
$ docker version
Client: Docker Engine - Community
 Version:           19.03.9
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        9d98839
 Built:             Fri May 15 00:34:28 2020
 OS/Arch:           linux/arm
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.9
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       9d98839
  Built:            Fri May 15 00:28:23 2020
  OS/Arch:          linux/arm
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

pi@white:~ $ docker pull alpine:3.9
3.9: Pulling from library/alpine
Digest: sha256:414e0518bb9228d35e4cd5165567fb91d26c6a214e9c95899e1e056fcd349011
Status: Image is up to date for alpine:3.9
docker.io/library/alpine:3.9
```

**Comment by kremerol** [comment]
Some ideas: 
  1. Because detecting the arm variant seams complex, what about "Giving the possibility to 'override' the way the variant is retrieve, for instance via a system variable (or a dockerd command line parameters)". This will permit to handle this kind of such 'strange' devices". That's not 'nice', not 'perfect', but that's also not really intrusive, and will for sure not break everything !
     For instance: CONTAINERD_ARM_VARIANT=v6
  2. "Set the arm variant by parsing the 'arch' command result instead of using /proc/cpuinfo".
  3. See if HWCAP could help (https://community.arm.com/developer/tools-software/oss-platforms/b/android-blog/posts/runtime-detection-of-cpu-features-on-an-armv8-a-cpu).

**Comment by thaJeztah** [comment]
Yes, the selection of the correct arm variant is not (yet) solved, partially due to the complex compatibility variations (e.g. armv7 machines being able to run both armv6 and armv7, but (IIRC) not in all cases), as well as some discussion that was ongoing about having a _generic_ selection algorithm (without deeply nesting very specific logic into the internals).

the main issue for tracking this is https://github.com/moby/moby/issues/34875 (which I see you commented on)

> Because detecting the arm variant seams complex, what about "Giving the possibility to 'override' the way the variant is retrieve, for instance via a system variable (or a dockerd command line parameters)". 

I think the `--platform` flag may be able to help with that. I think it's currently gated by the "experimental" option in the daemon (configure the daemon with `{"experimental":true}` in the `daemon.json`)

With that, you should be able to, e.g.;

```bash
docker run --platform=linux/arm/v6 alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME'
```

The current image store in dockerd does not support _storing_ multi-architecture images (so if if you pull different variants of the same image, then the latter will overwrite the former); this may be resolved at some point once the image store is integrated with the containerd.

**Comment by kremerol** [comment]
Hi @thaJeztah,
For me, there is the "selection" of the arm variant, and the "detection" of the arm variant:
1. First, you "detect" what is the current local arm architecture;
2. In a second step, you "select" the best available images according to the manifests.

Here, I am raising an issue about the "detection", which does not work on this specific ARM device. I agree that the "--platform" option could help to solve this "wrong detection" issue, but I don't like it too much because I don't know how to integrate it in docker-compose or K8S config. 
And about the "selection of the correct arm variant" (which is for me #34875), I think it is quite similar to amd64 vs 386: any amd64 system should be able to run 386 image, so what should do docker if there is only one 386 image defined in the manifest ?

**Comment by tianon** [comment]
This is a great summary of the (unfortunate) situation.  The platform matching in Docker has gotten a *lot* better than it used to be, but IMO the most important/relevant bit here (besides the suggestion of a "global" `--platform` flag on `dockerd`, which I'm 110% in favor of) is highlighting that kernel "bug" where a very popular set of ARMv6 devices report `CPU architecture: 7` (even though that's not accurate/correct).

What I'd propose (pending containerd maintainer opinions, obviously) is that if we detect the system to be `linux/arm/v7`, we then make one final check on `uname -m`, and if it is `armv6l`, we turn that down one notch to `linux/arm/v6` instead.

Is there also a containerd bug tracking this kernel mismatch? (I'm not aware of one, but I'd hate to have simply missed it, especially given there are *several* that are essentially tracking it for Docker :sweat_smile:)

**Comment by cpuguy83** [comment]
Sounds reasonable if it works for people.

**Comment by kremerol** [comment]
@tianon's proposal looks OK for me too (especially because, as him, I am not 110% in favor of the suggestion of a "global" --platform flag on dockerd).

**Comment by estesp** [comment]
I'm also OK with the proposal as it seems ARM CPU discovery mapping to variant has been a difficult thing to get "right" over the years--I guess my only question (which may not have an easy answer) is: is there any variant of the ARMv7 CPU family that might be adversely affected by this; e.g. might get "downgraded" to a v6 variant inadvertently?

**Comment by tianon** [comment]
This is technically "fixed" by https://github.com/containerd/containerd/pull/4530, although I don't think that made it into a containerd release yet?  Do we want to keep this (and related issues) open until it makes it into a release Docker's actually using?

**Comment by thaJeztah** [comment]
@tianon I see that PR was vendored in https://github.com/moby/moby/pull/41688, which is in Docker 20.10.0 and up; do you have a rPI0/rPI1 to verify if this is fixed with Docker 20.10?

**Comment by tianon** [comment]
I don't have a build of 20.10.0 handy, but I had some scripts handy that are able to build from https://github.com/docker/docker-ce, so I tested a recent commit there and it did work properly and appropriately pulled the `linux/arm/v5` image when I tried to run `debian:buster-slim` instead of pulling the `linux/arm/v7` version and then failing!

(Although amusingly, thanks to https://github.com/moby/moby/pull/41556, it included `WARNING: The requested image's platform (linux/arm) does not match the detected host platform (linux/arm/v6) and no specific platform was requested` which doesn't seem quite right, but harmless anyhow :sweat_smile:)

**Comment by ppseprus** [comment]
I am a little over my head hear, but I think I have a simliar/related issue while running a Docker build on an rpi B (arm32v6) model. 

```
Sending build context to Docker daemon  10.39MB
Step 1/11 : FROM arm32v6/node:14.15.4-alpine AS dependencies
14.15.4-alpine: Pulling from arm32v6/node
Digest: sha256:20cc61f7f6f2ec5754071c3ef2db622d29a785f66a5acc66a108f15b34621679
Status: Downloaded newer image for arm32v6/node:14.15.4-alpine
 ...
Step 5/11 : RUN npm install
 ---> [Warning] The requested image's platform (linux/arm) does not match the detected host platform (linux/arm/v6) and no specific platform was requested
 ---> Running in d97792defda3
```

Does anyone have any suggestions as for how to resolve this? Am I missing something? `npm install` has been running since  forever...

Thanks!

**Comment by jjlin** [comment]
> @tianon I see that PR was vendored in #41688, which is in Docker 20.10.0 and up; do you have a rPI0/rPI1 to verify if this is fixed with Docker 20.10?

I can confirm that this issue appears to be fixed in Docker 20.10.2 on a Raspberry Pi 1. I didn't test with 20.10.{0,1}, though.

**Comment by ondrejmo** [comment]
> @tianon I see that PR was vendored in #41688, which is in Docker 20.10.0 and up; do you have a rPI0/rPI1 to verify if this is fixed with Docker 20.10?

In Docker 20.10.1 on RPi0 it seems to be fixed as well.

**Comment by kremerol** [comment]
I did some very basic tests on newer docker versions on a RPI 1.2.
For me, the problem is not solved in docker 19.03.14, but it is solved with docker 20.10.0, 20.10.1 and 20.10.2:

```
sudo apt-get install docker-ce=5:19.03.14*
docker system prune -a -f && uname -m && docker run --rm alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME' || echo $?
=>
[...]
Status: Downloaded newer image for alpine:3.9
139 
```
But
```
sudo apt-get install docker-ce=5:20.10.0*
docker system prune -a -f && uname -m && docker run --rm alpine:3.9 sh -c 'cat /etc/*release | grep PRETTY_NAME' || echo $?
=>
[...]
Status: Downloaded newer image for alpine:3.9
WARNING: The requested image's platform (linux/arm) does not match the detected host platform (linux/arm/v6) and no specific platform was requested
PRETTY_NAME="Alpine Linux v3.9"
```
